pkg/hs/urbit-king/lib/Urbit/Vere/Ames.hs

```diff
@@ -106,13 +114,31 @@ ames'
   -> (Text -> RIO e ())
   -> RIO e ([Ev], RAcquire e (DriverApi NewtEf))
 ames' who isFake stderr = do
+  -- Unfortunately, we cannot use TBQueue because the only behavior
+  -- provided for when full is to block the writer, and we want to
+  -- instead drop the incoming packet on the floor. The implementation
+  -- below uses materially the same data structures as TBQueue, however.
   ventQ :: TQueue EvErr <- newTQueueIO
+  avail :: TVar Word <- newTVarIO queueBound
+  let
+    enqueuePacket p = do
+      vail <- readTVar avail
+      if vail > 0
+        then do
+          modifyTVar avail (subtract 1)
+          writeTQueue ventQ p
+        else pure ()  -- TODO debounced logging
```

Comment:
The idea here is to log whenever this condition is reached, debounced to at most once a second or something.

Two problems:
- How to log from STM? [most likely: unsafePerformIO :/]
- How to debounce? [assuming IO, http://hackage.haskell.org/package/auto-update-0.1.6/docs/Control-Debounce.html looks good and is by reputable author]

@jtobin, do you have any suggestion on best practices here?

Reply:
This will happen quite often on busy infrastructure nodes. I think it'd be better to duplicate vere's behavior here: count dropped packets and print every *N* (vere uses 1K). Storing a count of dropped packets is also useful, as it can be queried (see vere's SIGINFO/SIGUSR1 handlers on macos/linux, respectively).

Reply:
@pilfer-pandex Can you return a non-`()` value in `enqueuePacket` and log in IO conditionally on the result?  Toy example:

``` haskell
data PacketStatus =
    Enqueued
  | Debounced
  deriving Show

enqueuePacket foo = do
  vail <- readTVar foo
  if   vail > 0
  then do
    -- stuff
    pure Enqueued
  else do
    -- other stuff
    pure Debounced

queuePacketsThread = do
  foo    <- newTVarIO 0
  result <- atomically (enqueuePacket foo)
  print result
```

Reply:
+1 to something like @jtobin 's suggestion.

Reply:
I had thought that @jtobin's idea was impossible because the DriverApi passed to Serf has to be exactly an STM (), but I was totally mixing my components up and it is possible! Excellent suggestion, sir.

> SIGINFO/SIGUSR1

I looked this up, and how is this my first hearing of it? Pretty epic stuff. Of course ^T does nothing on kh today :/